google/cloud/sql/connector/rate_limiter.py

```diff
@@ -0,0 +1,114 @@
+"""
+Copyright 2021 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import asyncio
+import math
+import time
+from typing import Optional, Type
+from types import TracebackType
+
+
+class AsyncRateLimiter(object):
+    """
+    An asyncio-compatible rate limiter which uses the Token Bucket algorithm
+    (https://en.wikipedia.org/wiki/Token_bucket) to limit the number of function calls over a time interval using an event queue.
+
+    :type burst_size: int
+    :param: burst_size:
+        The max number of function calls that can occur at the same time
+        Default: 1
+
+    :type interval: int
+    :param: interval:
+        The period of time over which a number of calls equal to burst_size are allowed to complete. Default: 60s
+
+    :type queue_size: int
+    :param: interval:
+        The number of tasks that are allowed to be queued if rate limiter cannot be acquired. If exceeded, a QueueFull error will be thrown. A queue size of zero means the queue length has no upper bound. If setting queue size to a value greater than zero, queue_size must be greater than or equal to burst_size
+        Default: 0
+
+    :type loop: asyncio.AbstractEventLoop
+    :param: loop:
+        The event loop to use. If not provided, the default event loop will be used.
+
+
+    """
+
+    def __init__(
+        self,
+        burst_size: int = 1,
+        interval: float = 60,
+        loop: asyncio.AbstractEventLoop = None,
+    ) -> None:
+        self.interval = interval
+        self.burst_size = burst_size
+        self._loop = loop or asyncio.get_event_loop()
+        self._tokens = asyncio.BoundedSemaphore(burst_size, loop=self._loop)
+        self._last_token_update = time.time()
+
+    def _add_tokens(self) -> None:
+        """
+        Calculates how much time has passed since the last leak and removes the
+        appropriate amount of events from the queue.
+        Leaking is done lazily, meaning that if there is a large time gap between
+        leaks, the next set of calls might be a burst if burst_size > 1
+        """
+        time_elapsed = time.time() - self._last_token_update
+        amount = math.floor(time_elapsed * self.burst_size / self.interval)
```

Comment:
nit: We may want to add parenthesis here, I can't remember if python respects the order of operations

Reply:
I think the parentheses are unnecessary here. Python respects order of operations and even if it did, we're only multiplying and dividing so it doesn't matter. the result is the same regardless of which one you do first.